### PR TITLE
Block Editor: Use groups for InspectorControls

### PR DESCRIPTION
**Pull request by gziolo**
## Description

Related to #29247 where @youknowriad added support for groups in block toolbar.

The context for groups in `InspectorControls` can be found in the related comment https://github.com/WordPress/gutenberg/pull/34063#issuecomment-898280010:

> It's something I was talking about with @jasmussen in the context of absorbing the parent inspector controls as a follow-up for #33955. My initial idea was to have the following groups:
> - parent (new feature)
> - default (for the legacy features that exist in 3rd party blocks, but may also be a place for Block Styles picker and Block Transformation picker)
> - block (whatever the block defines that is specific to it)
> - advanced (it's basically `InspectorAdvancedControls`)
> 
> Now, there is a question about what we do about Global Styles UI. If that should be a single group or we should have more groups, for example:
> - default
> - layout
> - typography
> - border
> - color
> - dimensions

The initial implementation covers only two existing groups:
- `default` (`InspectorControls`)
- `advanced` (`InspectorAdvancedControls`)

## How has this been tested?

There should be no changes in how inspector controls work so I expect all e2e tests to pass.

I refactored `InspectorControls` and `InspectorAdvancedControls` in React Native implementation so it has to be confirmed it doesn't break anything.

## Types of changes

Refactoring. There should be no changes in how it works.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-editor/src/components/index.native.js [hunk 1]
```diff
@@ -19,7 +19,6 @@ export * from './gradients';
 export * from './font-sizes';
 export { AlignmentControl, AlignmentToolbar } from './alignment-control';
 export { default as InnerBlocks } from './inner-blocks';
-export { default as InspectorAdvancedControls } from './inspector-advanced-controls';
```
Comment (gziolo) [inline packages/block-editor/src/components/index.native.js:25]: I hope it wasn't used in React Native so far. I couldn't locate any reference in the code.

File: packages/block-editor/src/components/inspector-controls/fill.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	children,
+} ) {
+	const isDisplayed = useDisplayBlockControls();
+	const Fill = groups[ group ]?.Fill;
+	if ( ! Fill ) {
+		warning( `Unknown InspectorControl group "${ group }" provided.` );
+		return null;
+	}
+	if ( ! isDisplayed ) {
+		return false;
```
Comment (gziolo) [inline packages/block-editor/src/components/inspector-controls/fill.js]:
```suggestion
		return null;
```

**Comment by gziolo** [comment]
Given that #34063 is based on this branch, I plan to remove the slot for the group `block` and mark this PR as ready to go.

File: packages/block-editor/src/components/inspector-controls/README.md [hunk 3]
```diff
@@ -374,4 +179,44 @@ registerBlockType( 'my-plugin/inspector-controls-example', {
 } );
 ```
 
-{% end %}
+## InspectorAdvancedControls
```
Comment (gziolo) [inline packages/block-editor/src/components/inspector-controls/README.md:182]: Let's keep this section in the documentation until we mark the group as a stable API.

**Comment by jasmussen** [comment]
Thanks for working on this. From the instructions this appears to be mostly an under-the-hood change for the time being? Let me know if you need any mockup work for the absorption of parent inspector panels in child blocks.

**Comment by gziolo** [comment]
> Thanks for working on this. From the instructions this appears to be mostly an under-the-hood change for the time being? Let me know if you need any mockup work for the absorption of parent inspector panels in child blocks.

I want to start another PR to better channel discussion around explorations. This change will also let us exploring new groups for Global Styles related controls. 

And yes, this change is purely technical 👍🏻

**Comment by enejb** [comment]
I tested the PR on an iOS mobile build and everything worked for me as expected. So I think it is good to go from the mobile perspective.

**Comment by andrewserong** [comment]
This is looking great @gziolo, I like the addition of groups to InspectorControls and that the advanced controls just becomes another kind of group. I was wondering if it'd be worth adding a `constants.js` file and defining the `advanced` string once so that we avoid using string literals? Not a blocker, but I thought it might help avoid typos when we start adding in a number of other custom groups, so that we're referencing a canonical list of constants.

**Review by aaronrobertshaw — COMMENTED** [review]
This is great! Thanks @gziolo 👍 

The code looks good to me and the changes test well for both web and native.

I've also been testing this out by extending this PR to allow injection of controls into block support panels in https://github.com/WordPress/gutenberg/pull/34063, https://github.com/WordPress/gutenberg/pull/34157 and https://github.com/WordPress/gutenberg/pull/34065. This work has made achieving that much simpler and cleaner.

It might be best to get official approval on this from someone else that has more background on how this aligns with other APIs and the bigger picture. For me though, I'd be happy for this to land. 🚢 

>  I was wondering if it'd be worth adding a constants.js file and defining the advanced string once so that we avoid using string literals?

If we do this, I'd like to see it happen for the `BlockControls` groups as well.

**Comment by gziolo** [comment]
> > I was wondering if it'd be worth adding a constants.js file and defining the advanced string once so that we avoid using string literals?
> If we do this, I'd like to see it happen for the BlockControls groups as well.

@youknowriad, what do you think. I mostly followed your initial implementation. I don't think we use constants for props in other places. We can solve it with TypeScript types one day though.

**Comment by youknowriad** [comment]
> @youknowriad, what do you think. I mostly followed your initial implementation. I don't think we use constants for props in other places. We can solve it with TypeScript types one day though.

I wonder if the extra "import" to retrieve the consts would just add friction for block authors. Or is it just for internal usage?

**Comment by andrewserong** [comment]
> I wonder if the extra "import" to retrieve the consts would just add friction for block authors. Or is it just for internal usage?

I was imagining it as just for internal usage (mostly to avoid typos until we can use TypeScript types). Since the current approach here is consistent with the `BlockControls` group param, I think it'd be fine to leave it as is if you want, and we can look into possibly moving to consts in a separate PR if we'd like to do that.

**Comment by gziolo** [comment]
Let's move forward with this one and iterate further.

**Pull request merged**